File: src/agentshell/output.py
from __future__ import annotations

from dataclasses import dataclass


EXIT_OK = 0
EXIT_CHILD_FAILED = 1
EXIT_USAGE = 2
EXIT_RUNTIME = 3
EXIT_TIMEOUT = 124


@dataclass
class CommandResult:
    stdout: str
    stderr: str = ""
    exit_code: int = EXIT_OK


class AishError(Exception):
    def __init__(self, message: str, exit_code: int = EXIT_RUNTIME) -> None:
        super().__init__(message)
        self.message = message
        self.exit_code = exit_code


def join_lines(lines: list[str]) -> str:
    return "\n".join(line for line in lines if line is not None).rstrip() + "\n"


def truncate_value(value: str, limit: int = 180) -> str:
    if len(value) <= limit:
        return value
    return value[: max(0, limit - len("...truncated"))] + "...truncated"


def csv(items: list[str], empty: str = "-") -> str:
    return ",".join(items) if items else empty


def quote_command(args: list[str]) -> str:
    return " ".join(_quote_arg(arg) for arg in args)


def _quote_arg(arg: str) -> str:
    if not arg:
        return "''"
    if all(ch.isalnum() or ch in "-_./:=+" for ch in arg):
        return arg
    return "'" + arg.replace("'", "'\"'\"'") + "'"

File: src/output.js
export const EXIT_OK = 0;
export const EXIT_CHILD_FAILED = 1;
export const EXIT_USAGE = 2;
export const EXIT_RUNTIME = 3;
export const EXIT_TIMEOUT = 124;

export class AishError extends Error {
  constructor(message, exitCode = EXIT_RUNTIME) {
    super(message);
    this.exitCode = exitCode;
  }
}

export const result = (stdout = '', stderr = '', exitCode = EXIT_OK) => ({ stdout, stderr, exitCode });
export const joinLines = (lines) => `${lines.filter((line) => line != null).join('\n').trimEnd()}\n`;
export const csv = (items, empty = '-') => items.length ? items.join(',') : empty;
export const truncateValue = (value, limit = 180) => {
  const chars = Array.from(String(value));
  return chars.length <= limit ? String(value) : `${chars.slice(0, limit - 12).join('')}...truncated`;
};
export function quoteCommand(args) {
  return args.map((arg) => !arg ? "''" : /^[\w./:=+-]+$/u.test(arg) ? arg : `'${arg.replaceAll("'", `'"'"'`)}'`).join(' ');
}
